dlauthenticator/dlauthenticator.py

```diff
@@ -185,23 +213,129 @@ def authenticate(self, handler, data):
+    """
+    Google Cloud DataLab authenticator
+    """
+    DataLabAuthenticator.set_debug_user_path('/local/dlauth_debug_user')
+
+    def get_debug_user_info(self):
+        debug_user_info = {'username': None,
+                           'token': None}
+        if os.path.exists(self.DEBUG_USER_PATH):
+            with open(self.DEBUG_USER_PATH, 'r') as fd:
+                debug_user_info['username'] = fd.readline().strip()  # first line username
+                debug_user_info['token'] = fd.readline().strip()  # second line for token
+
+        return debug_user_info
+
+    def post_authenticate(self, handler, data, token):
+        # in debug mode, token might be None but a token
+        # might be passed down as part of the "data" structure
+        if token is None and "token" in data:
+            token = data['token']
+
+        # Extract user, uid, gid, hash, etc. from the token and return user_info.
+        user, uid, gid, hash = Util.split_auth_token(token)
+        self.log.info(f"Valid login for: {user}: {uid}: {gid}")
+        if not user or not uid or not gid:
+            self.log.warning(f"Login for user [{data['username']}] with token [{token}] failed")
+            return None
+
+        user_info = {
+            'name': data['username'],
+            'auth_state': {'uid': uid, 'gid': gid, 'token': token}
+        }
+        return user_info
+
+    @gen.coroutine
+    def pre_spawn_start(self, user, spawner):
+        # get_auth_state is a coroutine (async) wait for it with yield
+        # make sure the below is set in the config file
+        # c.Authenticator.enable_auth_state = True
+        # otherwise the below command will return None.
+        auth_state = yield spawner.user.get_auth_state()
+
+        # Dev note:
+        # print dir(self) or/and dir(spawner) to get attributes on these objects
+
+        user_options = spawner.user_options
+        selected_profile = user_options.get('profile')
+        self.log.info(f"user=[{spawner.user.name}] NB=[{selected_profile}]")
+
+        # Dev note:
+        # these get passed to the c.KubeSpawner.args.
+        # if for some reason we wanted make a logic decision and pass
+        # an argument to the spawner use, spawner.args.append:
+        # e.g.
+        # spawner.args.append('--LabApp.collaborative=True')
+
+        spawner.environment = {
+            # no need to set MEM_LIMIT here, it is already set in kubespawner_override
+            # and pass as env variable when the pod starts
+            # see kubespawner/spawner.py line 1861 make_pod ...
+            # 'MEM_LIMIT': str(c.KubeSpawner.mem_limit),
+            # 'MEM_LIMIT': 2147483648, # the big int value makes it crash
+            # 'MEM_LIMIT': '2147483648',
+            'UPSTREAM_TOKEN': auth_state['token'],
+            'NB_USER': spawner.user.name,
+            'NB_UID': auth_state['uid']
+        }
+
+
+class DevGCDataLabAuthenticator(GCDataLabAuthenticator):
```

Comment:
We could name this one something like `GCBaseAuthenticator` since theoretically it could be used anytime we want to use the jhub provided login pages. If it is something that should never be used in a production environment then leaving as is is probably preferred. Some other possible names are `GCAuthenticatorNative` or `GCAuthenticatorLocal` or `GCAuthenticatorNoRedirect` (naming things is hard)

Reply:
sounds reasonable, I think I'll use the GCAuthenticatorNoRedirect name as the GCBaseAuthenticator, to me, it implies a base class, which really isn't.
However I'll do this in a separate PR as the current production beta release uses this branch. ... duh ...